### PR TITLE
add autoincrement for run device

### PR DESCRIPTION
**Pull request by diitaz93**
## Description

This PR does 3 fixes to the database run tables: 

1. Modifies the primary keys of `run_device`, `run_metrics` and `sample_run_metrics` models in an alembic revision to include the autoincrement to the key values. For the primary keys to be modified, their use as foreign keys has to be temporarily suspended, that is why there is a drop constraint command before the modification and a creation of a constraint after.

2. The column `model` from `illumina_flow_cell` is set to nullable.

3. The column `run_metrics_id` in the `sample_run_metrics` didn't have its foreign key set in alembic, so now the constraint is added.

### How to prepare for test

- [x] Ssh to relevant server (depending on type of change)
- [x] Paxa the environment: `paxa`

### How to test

- [x] Do `alembic -c $ALEMBIC_CG_STAGE upgrade 5fd7e8758fb1`
- [x] Verify that the review worked 
- [x] Check that the tables `run_device`, `run_metrics` and `sample_run_metrics` have autoincrement 
- [x] Check that `sample_run_metrics` has `run_metrics.id` as the foreign key
- [x] Try adding a new row to each table
    - [x] Illumina flow cell with `model=None`
    - [x] Illumina sequencing metrics
    - [x] Illumina sample run metrics
- [x] Remove the entries
- [x]  alembic -c $ALEMBIC_CG_STAGE downgrade fe23de4ed528

## Review

- [x] Tests executed by SD
- [x] "Merge and deploy" approved by VJ, CO
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Deployed to stage:
```shell

```
- [ ] Deployed to production:
```shell

```

### CODE REVIEW

File: alembic/versions/5fd7e8758fb1_add_autoincrement_to_device_tables.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+def upgrade():
+    # Modify illumina flow cell primary key
+    op.drop_constraint(
+        constraint_name="illumina_flow_cell_ibfk_1",
+        table_name="illumina_flow_cell",
+        type_="foreignkey",
+    )
+    op.drop_constraint(constraint_name="fk_device_id", table_name="run_metrics", type_="foreignkey")
+    op.execute("ALTER TABLE run_device MODIFY COLUMN id INT NOT NULL AUTO_INCREMENT")
+    op.alter_column(
```
Comment (diitaz93) [inline alembic/versions/5fd7e8758fb1_add_autoincrement_to_device_tables.py:29]: the model column was set to not nullable originally, this changes it to nullable

File: alembic/versions/5fd7e8758fb1_add_autoincrement_to_device_tables.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    )
+    op.execute("ALTER TABLE run_metrics MODIFY COLUMN id INT NOT NULL AUTO_INCREMENT")
+    op.create_foreign_key(
+        constraint_name="illumina_sequencing_metrics_ibfk_1",
+        source_table="illumina_sequencing_metrics",
+        referent_table="run_metrics",
+        local_cols=["id"],
+        remote_cols=["id"],
+    )
+    op.create_foreign_key(
```
Comment (diitaz93) [inline alembic/versions/5fd7e8758fb1_add_autoincrement_to_device_tables.py:61]: this foreign key is necessary and was not included before

**Review by Vince-janv — APPROVED** [review]
Oof, a lot of strings to double check! I can't find anything wrong, but I think a proper test is needed before merging

**Review by ChrOertlin — APPROVED** [review]

**Comment by diitaz93** [comment]
## Tests
- Upgrading:
```
(cg) jsdiazboada@juasebmbp cg % alembic -c $ALEMBIC_CG_STAGE upgrade 5fd7e8758fb1  
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade fe23de4ed528 -> 5fd7e8758fb1, add_autoincrement_to_device_tables
(cg) jsdiazboada@juasebmbp cg %
```
Screenshot 2024-05-21 at 15 49 33

- Checking the autoincrement of tables
Screenshot 2024-05-21 at 15 51 45
Screenshot 2024-05-21 at 15 53 34
Screenshot 2024-05-21 at 15 54 22

- Check the new foreign key in `sample_run_metrics`
Screenshot 2024-05-21 at 15 56 32

**Comment by diitaz93** [comment]
## Tests
- Add IlluminaFlowCell with `model=None`
```python
new_illumina_flow_cell = IlluminaFlowCell(
    type=DeviceType.ILLUMINA,
    internal_id=fc.name,
    model=None
)
status_db.session.add(new_illumina_flow_cell)
status_db.session.commit()
```
Screenshot 2024-05-21 at 16 01 24

- Add Illumina sequencing metrics
```python
new_illumina_run_metrics = IlluminaRunMetrics(
        type=DeviceType.ILLUMINA,
        device_id=new_illumina_flow_cell.id,
        sequencer_type=fc.sequencer_type,
        sequencer_name=fc.sequencer_name,
        sequenced_at=fc.sequenced_at,
        data_availability=fc.status,
        archived_at=fc.archived_at,
        has_backup=fc.has_backup,
        total_reads=None,
        total_undetermined_reads=None,
        percent_q30=None,
        mean_quality_score=None,
        total_yield=None,
        yield_q30=None,
        cycles=None,
        demultiplexing_software=None,
        demultiplexing_software_version=None,
        sequencing_started_at=None,
        sequencing_completed_at=None,
        demultiplexing_started_at=None,
        demultiplexing_completed_at=None
    )
status_db.session.add(new_illumina_run_metrics)
status_db.session.commit()
```
Screenshot 2024-05-21 at 16 11 57

- Add Illumina sample run metrics
```python
new_illumina_sample_run_metrics = IlluminaSampleRunMetrics(
            sample_id=1,
            run_metrics_id=new_illumina_run_metrics.id,
            type=DeviceType.ILLUMINA,
            flow_cell_lane=1,
            total_reads_in_lane=10000,
            base_passing_q30_percent=93.3,
            base_mean_quality_score=35.27,
            created_at=None
        )
status_db.session.add(new_illumina_sample_run_metrics)
status_db.session.commit()
```
Screenshot 2024-05-21 at 16 17 14

**Comment by diitaz93** [comment]
## Test: downgrade
````
(cg) jsdiazboada@juasebmbp cg % alembic -c $ALEMBIC_CG_STAGE downgrade fe23de4ed528
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running downgrade 5fd7e8758fb1 -> fe23de4ed528, add_autoincrement_to_device_tables
(cg) jsdiazboada@juasebmbp cg % 
````

**Pull request merged**